Detect if entity_id should still be used for opensource-to-commerce and commerce-to-commerce migrations when Magento 2 was installed with Magento_CatalogStaging module disabled

dfelton opened the pull request:
### Description

Detects if `entity_id` should still be used during the Tier Prices step for opensource-to-commerce and commerce-to-commerce migrations that have had Magento installed with the Magento_CatalogStaging module disabled.

### Fixed Issues (if relevant)

When performing a migration from opensource-to-commerce or commerce-to-commerce, one has the option to have their `setup:install` step prior to migration performed with the following parameter:

```
--disable-modules=Magento_BundleStaging\
,Magento_CatalogImportExportStaging\
,Magento_CatalogInventoryStaging\
,Magento_CatalogStaging\
,Magento_CatalogStagingPageBuilder\
,Magento_CatalogUrlRewriteStaging\
,Magento_ConfigurableProductStaging\
,Magento_DownloadableStaging\
,Magento_GroupedProductStaging\
,Magento_ProductVideoStaging\
,Magento_WeeeStaging
```

Doing so, will initially cause the `Map Step` and `Tier Price` steps of the data-migration-tool to have errors regarding `row_id` not being mapped. However a developer has the freedom to copy the appropriate `*.xml.dist` files and modify them accordingly to resolve these issues.

During the Tier Price step however, we have code inside `Migration\Step\ConfigurablePrices\Helper::getDestinationFields()` that always assumes that the `row_id` column will be present if we are performing a migration where opensource is not the destination Magento platform.

This causes the following error during the Tier Price integrity check, that cannot be resolved with simply an XML configuration:

```
[2019-09-28 19:02:51][ERROR]: Destination fields are missing. Document: catalog_product_entity_decimal. Fields: row_id
```

### Manual testing scenarios

1. Install Magento 1.x EE
2. Install Magento Commerce 2.3.2 (with the modules listed in the description above disabled)
3. Copy the appropriate `config.xml.dist` file, modify the database credentials accordingly
4. Copy the corresponding `map-tier-price.xml.dist` file, and remove the `<move>` declarations from the `<source>` node, corresponding to `row_id` (There are two present for a EE 1.14.4.0 to Commerce 2.3.2 migration)
5. Update the `tier_price_map_file` XML node in the `config.xml` file from Step 3 above to reflect the new `map-tier-price.xml` from Step 4 above
6. Run `magento migrate:settings`
7. Run `magento migrate:data`

Following these steps alone, the data migration will still fail due to integrity check failures in the `Map Step`. However this Pull Request is not pertaining to the Map step. What *will* pass however, is the integrity check for the `Tier Price` step.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)

dfelton committed:
Detect if entity_id should still be used for opensource-to-commerce and commerce-to-commerce migrations when Magento 2 was installed with Magento_CatalogStaging module disabled.

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/signed)](https://cla.dev.magento.com/magento/data-migration-tool?pullRequest=722) <br/>All committers have signed the CLA.

dfelton commented:
(Force push just now was simply a revision of the author email on my commit)

dfelton committed:
Added ModuleListInterface to constructor docblock

dfelton commented:
Continuing to go through the efforts here to be able to truly get a migration accomplished with `Magento_CatalogStaging` disabled during the installation process. It appears there will be more work necessary on the PHP class side of things, to allow for a developer to perform a migration with this module disabled, and only have to modify appropriate XML configurations.

dfelton committed:
Corrects expectation of destination table structures when Magento 2 Commerce is installed with Magento_CatalogStaging module disabled.

Errors logged:
 - [ERROR]: Destination table structure does not meet expectation: catalog_category_entity_varchar
 - [ERROR]: Destination table structure does not meet expectation: catalog_product_entity_varchar

When migrating from commerce-to-commerce or opensource-to-commerce, the Helper was assuming that the Magento_CatalogStaging module had been installed during the initial setup:install command, and therefore assumed the need to replace `entity_id` with `row_id` for the two tables mentioned in the errors above.

dfelton committed:
Corrects value for $entityIdName if Magento 2 Commerce is installed with Magento_CatalogStaging disabled

With appropriate changes to the XML file for the Map step, we can get past the Integrity Check step on a commerce-to-commerce or opensource-to-opensource migration, when having installed Magento Commerce with Magento_CatalogStaging disabled. However despite a successful integrity check, the Map step would fail while in Data mode due to a hard coded assumption that row_id would be present in the catalog_product_entity_media_gallery_value_to_entity table for any migration that is not opensource-to-opensource.

Error Resolved: SQLSTATE[42S22]: Column not found: 1054 Unknown column 'row_id' in 'field list'

dfelton committed:
Resolves SQL errors when Maegnto 2 Commerce is installed with Magento_CatalogStaging disabled

Errors Resolved:

In Mysql.php line 110:

  SQLSTATE[42S22]: Column not found: 1054 Unknown column 'cpev.row_id' in 'on clause', query was: SELECT `cpev`.`value_id` FROM `catalog_product_entity_datetime` AS `cpev`
   INNER JOIN `catalog_product_entity` AS `cpe` ON cpe.entity_id = cpev.row_id WHERE (cpev.value IS NOT NULL) AND (cpev.attribute_id NOT IN (SELECT `eea`.`attribute_id` FROM `eav_entity_attribute` AS `eea` WHERE (eea.attribute_set_id = cpe.attribute_s
  et_id)))

In Mysql.php line 91:

  SQLSTATE[42S22]: Column not found: 1054 Unknown column 'cpev.row_id' in 'on clause'

dfelton commented:
With the latest push (along with the appropriate changes to various `*.xml.dist` files)  I have now been able to perform a `commerce-to-commerce` migration with Magento EE 1.14.4.0 to Magento Commerce 2.3.2, with the modules in the PR's original description disabled during the `setup:install` command prior to.

Steps skipped due to not being applicable to my current project:

- Log Step
- Ratings Step
- VisualMerchandiser Step

The migration, to the best of my knowledge, has been successful. 

Further review shows issues with the `catalogsearch_fulltext` indexer post-migration. This however is not related to the data-migration-tool, and strictly an issue with Magento Commerce 2.3.2. This was confirmed by installing Magento Commerce 2.3.2 (with above modules disabled during install step), adding a product, and running the indexer. Because this issue pertains to Magento Commerce, I will report it through my account on Magento.com rather than further discussing here.

Please do let me know if this PR is acceptable, or if further changes are desired. My primary goal here is to be able to allow developers to be able to perform a migration under the following scenario:

- A stock copy of Magento (EE or CE) 1.x to a stock copy of Magento 2.x Commerce
- Have module(s) disabled during the `setup:install` of Magento 2 Commerce if desired
- Only have to modify `*.xml.dist` files rather than write plugins and/or extend `Migration\*` classes to achieve a successful migration.

Thank you,
Darren Felton

victor-v-rad commented:
Hi @dfelton 
Sorry for the late reply. Internal ticket MC-21538 to process it

dfelton commented:
Thanks @victor-v-rad. No worries here.

If there's any help I can be let me know. For now I've put down any effort to take this further due to the indexer issue in Magento Commerce with said modules disabled.

If it of any help, I submitted support ticket 248120 to on Magento.com to report the indexer issues there.

ghost commented:
Hi @dfelton, thank you for your contribution! </br>Please, complete [Contribution Survey](https://opensource.magento.com/survey/contribution/magento+data-migration-tool+722), it will take less than a minute. </br>Your feedback will help us to improve contribution process.

victor-v-rad commented:
Thank you @dfelton for your great job on this issue. I created https://github.com/magento/data-migration-tool/issues/728 with reference to this pull request. So the people who have similar issue now will see the information you shared.

dfelton commented:
>  For now I've put down any effort to take this further due to the indexer issue in Magento Commerce with said modules disabled.

Quick note that I have confirmed that said issue is resolved in Magento Commerce 2.3.3. 

Steps Taken:

1 - Install Magento Commerce 2.3.3 with the following CLI command:

    ./bin/magento setup:install \
    --disable-modules=Magento_BundleStaging,Magento_CatalogImportExportStaging,Magento_CatalogInventoryStaging,Magento_CatalogStaging,Magento_CatalogStagingPageBuilder,Magento_CatalogUrlRewriteStaging,Magento_ConfigurableProductStaging,Magento_DownloadableStaging,Magento_GroupedProductStaging,Magento_ProductVideoStaging,Magento_WeeeStaging \
    --base-url=http://magento.commerce.2.3.3.localhost/ \
    --db-name=magento_commerce_2_3_3 \
    --backend-frontname=admin

2. Add Admin user via CLI
3. Compile static content
4. Compile DI
5. Log into admin
6. Create a new product
7. Execute ./bin/magento indexer:reindex catalogsearch_fulltext
    - (Command completed successfully)

For good measure I performed a full reindex, which completed successfully.
Lastly, I reconfirmed modules were installed disabled using the module:status command:

    List of disabled modules:
    Magento_CatalogImportExportStaging
    Magento_CatalogStaging
    Magento_GroupedProductStaging
    Magento_BundleStaging
    Magento_ProductVideoStaging
    Magento_ConfigurableProductStaging
    Magento_CatalogStagingPageBuilder
    Magento_DownloadableStaging
    Magento_CatalogUrlRewriteStaging
    Magento_CatalogInventoryStaging
    Magento_WeeeStaging
